docs/rfcs/20220131_raft_snapshots.md

```diff
@@ -0,0 +1,150 @@
+- Feature Name: Raft snapshots
+- Status: draft
+- Start Date: 2022-01-31
+- Authors: Denis Rystsov
+- Issue: TBD
+
+# Executive Summary
+
+Introduce Raft snapshots not bound to log truncation in order to solve compaction-related problems with idempotency, transactions, and offset translator, and pave the way to custom state machines.
+
+## What is being proposed
+
+Make the snapshot before the compaction and ship it to all (or the majority of) the nodes in a cluster so that each node sees the same aggregate of the log. Currently, when Redpanda does compaction, it introduces holes to the log and when a follower replays the log to build an aggregate, it ends up with different view.
+
+# Motivation
+## Why are we doing this?
+
+To fix existing problems that prevent transactions from working with the compacted topics. To introduce infrastructure that supports custom (probably WASM based) state machines.
+
+## What use cases does it support?
+
+ - Idempotency
+ - Transactions
+ - Offset translator
+ - Custom state machines
+
+# Guide-level explanation
+## Context
+
+Some of the Redpanda operations require knowledge about all of the records that are passed through the partition (Raft log) up to the current moment (let's call them aggregate-dependent).
```

Comment:
Thinking out loud to help my understanding here: Naming is hard. Nit: maybe "aggregation-dependent" since it is the process (aggregation) and not the result (aggregate) that these operations depend on. I'm thinking of this concept as "non-trivially compactable".
If we name Kafka's idea of compaction "trivial key-based compaction", we are just noting that this is not the same as STM or transaction snapshotting in general; the snapshot needs to reflect the result of the applied operations. Simply eliding older operations with the same key is not sufficient for the more general case of STMs/transactions/operations.